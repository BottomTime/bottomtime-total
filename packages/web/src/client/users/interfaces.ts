import { ProfileVisibility, UserRole } from '@/constants';
import { z } from 'zod';

export const ProfileCertificationDataSchema = z.object({
  agency: z.string().optional(),
  course: z.string(),
  date: z.coerce.date().optional(),
});
export type ProfileCertificationData = z.infer<
  typeof ProfileCertificationDataSchema
>;

export const ProfileDataSchema = z.object({
  avatar: z.string().optional(),
  bio: z.string().optional(),
  birthdate: z.string().optional(),
  customData: z.record(z.string(), z.unknown()).optional(),
  certifications: z.array(ProfileCertificationDataSchema).optional(),
  experienceLevel: z.string().optional(),
  location: z.string().optional(),
  name: z.string().optional(),
  profileVisibility: z.nativeEnum(ProfileVisibility),
  startedDiving: z.string().optional(),
});
export type ProfileData = z.infer<typeof ProfileDataSchema>;

export const UserDataSchema = z.object({
  id: z.string(),
  email: z.string().optional(),
  emailVerified: z.coerce.boolean(),
  hasPassword: z.coerce.boolean(),
  isLockedOut: z.coerce.boolean(),
  lastLogin: z.coerce.date().optional(),
  lastPasswordChange: z.coerce.date().optional(),
  memberSince: z.coerce.date(),
  role: z.nativeEnum(UserRole),
  username: z.string(),

  profile: ProfileDataSchema,
});
export type UserData = z.infer<typeof UserDataSchema>;

export interface Profile extends ProfileData {
  readonly userId: string;
  readonly username: string;
  readonly memberSince: Date;

  save(): Promise<void>;
  toJSON(): object;
}

export interface User extends Readonly<UserData> {
  readonly profile: Profile;

  changeEmail(newEmail: string): Promise<void>;
  changePassword(oldPassword: string, newPassword: string): Promise<void>;
  changeUsername(newUsername: string): Promise<void>;
  requestVerificationEmail(): Promise<void>;
  verifyEmail(token: string): Promise<boolean>;

  toJSON(): UserData;
}

export interface CreateUserOptions {
  username: string;
  password?: string;
  email?: string;
  profile?: ProfileData;
}

export interface UserManager {
  authenticateUser(usernameOrEmail: string, password: string): Promise<User>;
  createUser(options: CreateUserOptions): Promise<User>;
  getCurrentUser(): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User>;
  isUsernameOrEmailAvailable(usernameOrEmail: string): Promise<boolean>;
  requestPasswordReset(usernameOrEmail: string): Promise<void>;
  resetPassword(
    username: string,
    token: string,
    newPassword: string,
  ): Promise<boolean>;
}
