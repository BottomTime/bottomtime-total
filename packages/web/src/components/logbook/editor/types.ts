import {
  DepthUnit,
  DiveSiteDTO,
  ExposureSuit,
  LogEntryDTO,
  OperatorDTO,
  PressureUnit,
  TankDTO,
  TankMaterial,
  TemperatureUnit,
  TrimCorrectness,
  WeightCorrectness,
  WeightUnit,
} from '@bottomtime/api';

import dayjs from 'dayjs';
import { v7 as uuid } from 'uuid';
import { computed } from 'vue';

import { SelectOption } from '../../../common';

export interface LogEntryBasicInfo {
  avgDepth: number | string;
  bottomTime: string | number;
  depthUnit: DepthUnit;
  duration: string | number;
  entryTime?: Date;
  entryTimezone: string;
  logNumber: string | number;
  maxDepth: number | string;
  surfaceInterval: string | number;
  tags: string[];
}

export interface LogEntryConditions {
  airTemp: number | string;
  waterTemp: number | string;
  thermocline: number | string;
  tempUnit: TemperatureUnit;

  chop: number;
  current: number;
  visibility: number;

  weather: string;
}

export interface LogEntryEquipment {
  weight: number | string;
  weightUnit: WeightUnit;
  weightCorrectness: WeightCorrectness | '';
  trim: TrimCorrectness | '';
  exposureSuit: ExposureSuit | '';
  boots?: boolean;
  camera?: boolean;
  hood?: boolean;
  gloves?: boolean;
  scooter?: boolean;
  torch?: boolean;
}

export interface LogEntryAir {
  id: string;
  tankId: string;
  tankInfo?: TankDTO;
  doubles: boolean;
  startPressure: string | number;
  endPressure: string | number;
  pressureUnit: PressureUnit;
  hePercent: string | number;
  o2Percent: string | number;
}

export interface LogEntryLocation {
  site?: DiveSiteDTO;
  operator?: OperatorDTO;
}

export interface LogEntryNotes {
  notes: string;
  rating?: number;
}

export interface LogEntryFormData {
  air: LogEntryAir[];
  basicInfo: LogEntryBasicInfo;
  conditions: LogEntryConditions;
  equipment: LogEntryEquipment;
  location: LogEntryLocation;
  notes: LogEntryNotes;
}

export const Timezones = computed<SelectOption[]>(() =>
  Intl.supportedValuesOf('timeZone').map((tz) => ({
    label: tz,
    value: tz,
  })),
);

export const BlankAirEntry: LogEntryAir = {
  id: '',
  startPressure: '',
  endPressure: '',
  doubles: false,
  pressureUnit: PressureUnit.Bar,
  hePercent: '',
  o2Percent: '',
  tankId: '',
} as const;

function dtoToAirFormData(
  entry: LogEntryDTO,
  defaultPressureUnit: PressureUnit,
): LogEntryAir[] {
  if (!entry.air || entry.air.length === 0) {
    return [
      {
        ...BlankAirEntry,
        id: uuid(),
        pressureUnit: defaultPressureUnit,
      },
    ];
  }

  return entry.air.map((air) => ({
    doubles: air.count > 1,
    endPressure: air.endPressure,
    hePercent: air.hePercent ?? '',
    id: uuid(),
    o2Percent: air.o2Percent ?? '',
    pressureUnit: air.pressureUnit,
    startPressure: air.startPressure,
    tankId: '',
    tankInfo: air.name
      ? {
          id: '',
          material: air.material,
          name: air.name,
          volume: air.volume,
          isSystem: false,
          workingPressure: air.workingPressure,
        }
      : undefined,
  }));
}

type DefaultUnitOptions = {
  depth: DepthUnit;
  pressure: PressureUnit;
  temperature: TemperatureUnit;
  weight: WeightUnit;
};

export function dtoToFormData(
  entry: LogEntryDTO,
  defaultUnits: DefaultUnitOptions,
): LogEntryFormData {
  return {
    basicInfo: {
      avgDepth: entry.depths?.averageDepth || '',
      bottomTime: entry.timing.bottomTime || '',
      depthUnit: entry.depths?.depthUnit || defaultUnits.depth,
      duration: entry.timing.duration === -1 ? '' : entry.timing.duration,
      entryTimezone: entry.timing.timezone || dayjs.tz.guess(),
      logNumber: entry.logNumber || '',
      maxDepth: entry.depths?.maxDepth || '',
      surfaceInterval: '', // TODO
      entryTime: Number.isNaN(entry.timing.entryTime)
        ? undefined
        : new Date(entry.timing.entryTime),
      tags: entry.tags ?? [],
    },

    conditions: {
      airTemp: entry.conditions?.airTemperature || '',
      waterTemp: entry.conditions?.surfaceTemperature || '',
      thermocline: entry.conditions?.bottomTemperature || '',
      tempUnit: entry.conditions?.temperatureUnit || defaultUnits.temperature,

      chop: entry.conditions?.chop || 0,
      current: entry.conditions?.current || 0,
      visibility: entry.conditions?.visibility || 0,

      weather: entry.conditions?.weather || '',
    },

    equipment: {
      weight: entry.equipment?.weight || '',
      weightUnit: entry.equipment?.weightUnit || defaultUnits.weight,
      weightCorrectness: entry.equipment?.weightCorrectness || '',
      trim: entry.equipment?.trimCorrectness || '',
      exposureSuit: entry.equipment?.exposureSuit || '',

      boots: entry.equipment?.boots,
      camera: entry.equipment?.camera,
      gloves: entry.equipment?.gloves,
      hood: entry.equipment?.hood,
      scooter: entry.equipment?.scooter,
      torch: entry.equipment?.torch,
    },

    location: {},

    notes: {
      notes: entry.notes || '',
      rating: entry.rating,
    },

    air: dtoToAirFormData(entry, defaultUnits.pressure),
  };
}

function getNumericValue(
  value: string | number | undefined,
): number | undefined {
  return typeof value === 'number' ? value : undefined;
}

function getNullableEnumValue<T>(value: T | ''): T | undefined {
  return value === '' ? undefined : value;
}

export function formDataToDTO(
  original: LogEntryDTO,
  data: LogEntryFormData,
): LogEntryDTO {
  return {
    ...original,
    air: data.air.length
      ? data.air.map((air) => ({
          count: air.doubles ? 2 : 1,
          endPressure: getNumericValue(air.endPressure) ?? 0,
          material: air.tankInfo?.material ?? TankMaterial.Aluminum,
          name: air.tankInfo?.name || '',
          volume: air.tankInfo?.volume ?? 0,
          workingPressure: air.tankInfo?.workingPressure ?? 0,
          hePercent: getNumericValue(air.hePercent),
          o2Percent: getNumericValue(air.o2Percent),
          pressureUnit: air.pressureUnit,
          startPressure: getNumericValue(air.startPressure) ?? 0,
        }))
      : undefined,
    conditions: {
      airTemperature: getNumericValue(data.conditions.airTemp),
      bottomTemperature: getNumericValue(data.conditions.thermocline),
      chop: data.conditions.chop,
      current: data.conditions.current,
      surfaceTemperature: getNumericValue(data.conditions.waterTemp),
      temperatureUnit: data.conditions.tempUnit,
      visibility: data.conditions.visibility,
      weather: data.conditions.weather,
    },
    depths: {
      averageDepth: getNumericValue(data.basicInfo.avgDepth),
      depthUnit: data.basicInfo.depthUnit,
      maxDepth: getNumericValue(data.basicInfo.maxDepth),
    },
    equipment: {
      boots: data.equipment.boots,
      camera: data.equipment.camera,
      exposureSuit: getNullableEnumValue(data.equipment.exposureSuit),
      gloves: data.equipment.gloves,
      hood: data.equipment.hood,
      scooter: data.equipment.scooter,
      torch: data.equipment.torch,
      trimCorrectness: getNullableEnumValue(data.equipment.trim),
      weight: getNumericValue(data.equipment.weight),
      weightCorrectness: getNullableEnumValue(data.equipment.weightCorrectness),
      weightUnit: data.equipment.weightUnit,
    },
    timing: {
      duration: getNumericValue(data.basicInfo.duration) ?? 0,
      entryTime: data.basicInfo.entryTime?.valueOf() ?? 0,
      timezone: data.basicInfo.entryTimezone,
      bottomTime: getNumericValue(data.basicInfo.bottomTime),
    },
    logNumber: getNumericValue(data.basicInfo.logNumber),
    notes: data.notes.notes,
    operator: data.location.operator,
    site: data.location.site,
    rating: data.notes.rating,
    tags: data.basicInfo.tags.length ? data.basicInfo.tags : undefined,
  };
}
