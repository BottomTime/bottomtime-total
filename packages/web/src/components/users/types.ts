import { NotificationDTO } from '@bottomtime/api';

export type NotificationWithSelection = NotificationDTO & { selected: boolean };
