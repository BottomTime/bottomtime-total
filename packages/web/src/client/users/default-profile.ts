import { assertValid } from '@/helpers';
import { SuperAgentStatic } from 'superagent';
import {
  Profile,
  ProfileCertificationData,
  ProfileData,
  ProfileDataSchema,
  UserData,
} from './interfaces';
import { ProfileVisibility } from '@/constants';

export class DefaultProfile implements Profile {
  private static readonly JsonOmitKeys = new Set([
    'userId',
    'memberSince',
    'username',
  ]);

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get userId(): string {
    return this.data.id;
  }
  get memberSince(): Date {
    return this.data.memberSince;
  }
  get username(): string {
    return this.data.username;
  }

  get avatar(): string | undefined {
    return this.data.profile.avatar;
  }
  set avatar(value: string | undefined) {
    this.data.profile.avatar = value;
  }

  get bio(): string | undefined {
    return this.data.profile.bio;
  }
  set bio(value: string | undefined) {
    this.data.profile.bio = value;
  }

  get birthdate(): string | undefined {
    return this.data.profile.birthdate;
  }
  set birthdate(value: string | undefined) {
    this.data.profile.birthdate = value;
  }

  get customData(): Record<string, unknown> | undefined {
    return this.data.profile.customData;
  }
  set customData(value: Record<string, unknown> | undefined) {
    this.data.profile.customData = value;
  }

  get certifications(): ProfileCertificationData[] | undefined {
    return this.data.profile.certifications;
  }
  set certifications(value: ProfileCertificationData[] | undefined) {
    this.data.profile.certifications = value;
  }

  get experienceLevel(): string | undefined {
    return this.data.profile.experienceLevel;
  }
  set experienceLevel(value: string | undefined) {
    this.data.profile.experienceLevel = value;
  }

  get location(): string | undefined {
    return this.data.profile.location;
  }
  set location(value: string | undefined) {
    this.data.profile.location = value;
  }

  get name(): string | undefined {
    return this.data.profile.name;
  }
  set name(value: string | undefined) {
    this.data.profile.name = value;
  }

  get profileVisibility(): ProfileVisibility {
    return this.data.profile.profileVisibility;
  }
  set profileVisibility(value: ProfileVisibility) {
    this.data.profile.profileVisibility = value;
  }

  get startedDiving(): string | undefined {
    return this.data.profile.startedDiving;
  }
  set startedDiving(value: string | undefined) {
    this.data.profile.startedDiving = value;
  }

  async save(): Promise<void> {
    const profile = assertValid<ProfileData>(
      this.data.profile,
      ProfileDataSchema,
    );
    await this.agent.put(`/api/profiles/${this.username}`).send(profile);
    this.data.profile = profile;
  }

  toJSON(): ProfileData {
    return assertValid(this.data.profile, ProfileDataSchema);
  }
}
