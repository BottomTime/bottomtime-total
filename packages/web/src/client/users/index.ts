export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
