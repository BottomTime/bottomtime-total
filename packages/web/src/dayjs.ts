import dayjs from 'dayjs';
import localizedFormat from 'dayjs/plugin/localizedFormat';
import relativeTime from 'dayjs/plugin/relativeTime';
import tz from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';

dayjs.extend(localizedFormat);
dayjs.extend(relativeTime);
dayjs.extend(tz);
dayjs.extend(utc);

export default dayjs;
