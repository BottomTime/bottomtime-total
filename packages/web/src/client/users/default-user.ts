import { SuperAgentStatic } from 'superagent';
import { DefaultProfile } from './default-profile';
import { Profile, User, UserData, UserDataSchema } from './interfaces';
import { UserRole } from '@/constants';
import { assertValid } from '@/helpers';

export class DefaultUser implements User {
  private _profile: Profile | undefined;

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get profile(): Profile {
    if (!this._profile) {
      this._profile = new DefaultProfile(this.agent, this.data);
    }
    return this._profile;
  }

  get id(): string {
    return this.data.id;
  }

  get email(): string | undefined {
    return this.data.email;
  }

  get emailVerified(): boolean {
    return this.data.emailVerified;
  }

  get hasPassword(): boolean {
    return this.data.hasPassword;
  }

  get isLockedOut(): boolean {
    return this.data.isLockedOut;
  }

  get lastLogin(): Date | undefined {
    return this.data.lastLogin;
  }

  get lastPasswordChange(): Date | undefined {
    return this.data.lastPasswordChange;
  }

  get memberSince(): Date {
    return this.data.memberSince;
  }

  get role(): UserRole {
    return this.data.role;
  }

  get username(): string {
    return this.data.username;
  }

  async changeEmail(newEmail: string): Promise<void> {
    await this.agent
      .post(`/api/users/${this.data.username}/changeEmail`)
      .send({ newEmail });

    Object.assign(this.data, {
      email: newEmail,
      emailVerified: false,
    });
  }

  async changePassword(
    oldPassword: string,
    newPassword: string,
  ): Promise<void> {
    await this.agent
      .post(`/api/users/${this.data.username}/changePassword`)
      .send({ oldPassword, newPassword });
  }

  async changeUsername(newUsername: string): Promise<void> {
    await this.agent
      .post(`/api/users/${this.data.username}/changeUsername`)
      .send({ newUsername });

    Object.assign(this.data, {
      username: newUsername,
    });
  }

  async requestVerificationEmail(): Promise<void> {
    await this.agent.post(
      `/api/users/${this.data.username}/requestEmailVerification`,
    );
  }

  async verifyEmail(token: string): Promise<boolean> {
    const {
      body: { verified },
    } = await this.agent
      .post(`/api/users/${this.data.username}/verifyEmail`)
      .send({ token });
    return verified === true;
  }

  toJSON(): UserData {
    return assertValid(this.data, UserDataSchema);
  }
}
