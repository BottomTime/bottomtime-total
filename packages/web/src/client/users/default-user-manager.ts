import request from 'superagent';
import { SuperAgentStatic } from 'superagent';
import { z } from 'zod';

import { DefaultUser } from './default-user';
import {
  CreateUserOptions,
  User,
  UserData,
  UserDataSchema,
  UserManager,
} from './interfaces';
import { assertValid } from '@/helpers';

const ResetPasswordResponseSchema = z.object({
  succeeded: z.coerce.boolean(),
});
type ResetPasswordResponse = z.infer<typeof ResetPasswordResponseSchema>;

export class DefaultUserManager implements UserManager {
  constructor(private readonly agent: SuperAgentStatic) {}

  async authenticateUser(
    usernameOrEmail: string,
    password: string,
  ): Promise<User> {
    const { body } = await this.agent.post('/api/auth/login').send({
      usernameOrEmail,
      password,
    });
    const userData = assertValid<UserData>(body, UserDataSchema);
    return new DefaultUser(this.agent, userData);
  }

  async createUser(options: CreateUserOptions): Promise<User> {
    const { body } = await this.agent
      .put(`/api/users/${options.username}`)
      .send({
        email: options.email,
        password: options.password,
        profile: options.profile,
      });

    const userData = assertValid(body, UserDataSchema);
    return new DefaultUser(this.agent, userData);
  }

  async getCurrentUser(): Promise<User | undefined> {
    const { body } = await this.agent.get('/api/auth/me');
    if (body.anonymous) {
      return undefined;
    }

    const userData = assertValid(body, UserDataSchema);
    return new DefaultUser(this.agent, userData);
  }

  async getUserByUsername(username: string): Promise<User> {
    const { body } = await this.agent.get(`/api/users/${username}`);
    const userData = assertValid(body, UserDataSchema);
    return new DefaultUser(this.agent, userData);
  }

  isUsernameOrEmailAvailable(usernameOrEmail: string): Promise<boolean> {
    return new Promise((resolve, reject) => {
      // Why the hell does Superagent not support HEAD requests!?
      request
        .head(`/api/users/${usernameOrEmail}`)
        .then(() => {
          resolve(false);
        })
        .catch((error) => {
          if (error.status === 404) resolve(true);
          else reject(error);
        });
    });
  }

  async requestPasswordReset(usernameOrEmail: string): Promise<void> {
    await this.agent.post(`/api/users/${usernameOrEmail}/requestPasswordReset`);
  }

  async resetPassword(
    username: string,
    token: string,
    newPassword: string,
  ): Promise<boolean> {
    const { body } = await this.agent
      .post(`/api/users/${username}/`)
      .send({ token, newPassword });

    const { succeeded } = assertValid<ResetPasswordResponse>(
      body,
      ResetPasswordResponseSchema,
    );
    return succeeded;
  }
}
