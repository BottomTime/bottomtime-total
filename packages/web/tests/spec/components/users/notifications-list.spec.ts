import { ApiList, ListNotificationsResponseSchema } from '@bottomtime/api';

import { ComponentMountingOptions, mount } from '@vue/test-utils';

import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

import NotificationsListItem from '../../../../src/components/users/notifications-list-item.vue';
import NotificationsList from '../../../../src/components/users/notifications-list.vue';
import { NotificationWithSelection } from '../../../../src/components/users/types';
import TestData from '../../../fixtures/notifications.json';

dayjs.extend(relativeTime);

const Counts = '[data-testid="notification-counts"]';
const LoadMoreButton = '[data-testid="btn-load-more"]';
const NoNotificationsMessage = '[data-testid="msg-no-notifications"]';

const SelectAll = '[data-testid="btn-select-all"]';
const SelectNone = '[data-testid="btn-select-none"]';

const BulkDismiss = '[data-testid="btn-bulk-dismiss"]';
const BulkUndismiss = '[data-testid="btn-bulk-undismiss"]';
const BulkDelete = '[data-testid="btn-bulk-delete"]';

describe('NotificationsList component', () => {
  let opts: ComponentMountingOptions<typeof NotificationsList>;
  let testData: ApiList<NotificationWithSelection>;

  beforeAll(() => {
    const parsed = ListNotificationsResponseSchema.parse(TestData);
    testData = {
      data: parsed.data.map((n) => ({ ...n, selected: false })),
      totalCount: parsed.totalCount,
    };
  });

  beforeEach(() => {
    opts = {
      props: {
        notifications: testData,
      },
    };
  });

  it('will render a list of notifications', () => {
    const wrapper = mount(NotificationsList, opts);
    expect(wrapper.get(Counts).text()).toMatchSnapshot();

    const items = wrapper.findAllComponents(NotificationsListItem);
    expect(items).toHaveLength(50);
    items.forEach((item, i) => {
      expect(item.props('notification')).toEqual(testData.data[i]);
    });
    expect(wrapper.get(LoadMoreButton).isVisible()).toBe(true);
    expect(wrapper.find(NoNotificationsMessage).exists()).toBe(false);
  });

  it('will render a message if no notifications are available', async () => {
    const wrapper = mount(NotificationsList, opts);
    await wrapper.setProps({ notifications: { data: [], totalCount: 0 } });

    expect(wrapper.find(LoadMoreButton).exists()).toBe(false);
    expect(wrapper.find(NoNotificationsMessage).isVisible()).toBe(true);

    const items = wrapper.findAllComponents(NotificationsListItem);
    expect(items).toHaveLength(0);
  });

  it('will emit event if load more button is clicked', async () => {
    const wrapper = mount(NotificationsList, opts);
    await wrapper.get(LoadMoreButton).trigger('click');
    expect(wrapper.emitted('load-more')).toBeDefined();
  });

  it('will bubble up select event', async () => {
    const wrapper = mount(NotificationsList, opts);
    const item = wrapper.getComponent(NotificationsListItem);
    await item.vm.$emit('select', item.props('notification'), true);
    await item.vm.$emit('select', item.props('notification'), false);

    expect(wrapper.emitted('select')).toEqual([
      [item.props('notification'), true],
      [item.props('notification'), false],
    ]);
  });

  it('will bubble up delete event', async () => {
    const wrapper = mount(NotificationsList, opts);
    const item = wrapper.getComponent(NotificationsListItem);
    await item.vm.$emit('delete', item.props('notification'));
    expect(wrapper.emitted('delete')).toEqual([[item.props('notification')]]);
  });

  it('will bubble up dismiss event', async () => {
    const wrapper = mount(NotificationsList, opts);
    const item = wrapper.getComponent(NotificationsListItem);
    await item.vm.$emit('toggle-dismiss', item.props('notification'));
    expect(wrapper.emitted('dismiss')).toEqual([[item.props('notification')]]);
  });

  it('will bubble up undismiss event', async () => {
    const wrapper = mount(NotificationsList, opts);
    await wrapper.setProps({
      notifications: {
        data: [{ ...testData.data[0], dismissed: true }],
        totalCount: 1,
      },
    });
    const item = wrapper.getComponent(NotificationsListItem);
    await item.vm.$emit('toggle-dismiss', item.props('notification'));
    expect(wrapper.emitted('undismiss')).toEqual([
      [item.props('notification')],
    ]);
  });

  it('will select all notifications', async () => {
    const wrapper = mount(NotificationsList, opts);
    await wrapper.get(SelectAll).trigger('click');
    expect(wrapper.emitted('select')).toEqual([[testData.data, true]]);
  });

  it('will unselect all notifications', async () => {
    const wrapper = mount(NotificationsList, opts);
    await wrapper.get(SelectNone).trigger('click');
    expect(wrapper.emitted('select')).toEqual([[testData.data, false]]);
  });

  describe('when performing bulk operations', () => {
    let selectedNotifications: NotificationWithSelection[];
    let testDataWithSelections: ApiList<NotificationWithSelection>;

    beforeAll(() => {
      testDataWithSelections = {
        data: testData.data.map((n) => ({ ...n })),
        totalCount: testData.totalCount,
      };
      selectedNotifications = [
        testDataWithSelections.data[0],
        testDataWithSelections.data[7],
        testDataWithSelections.data[12],
        testDataWithSelections.data[23],
      ];
      testDataWithSelections.data[0].selected = true;
      testDataWithSelections.data[7].selected = true;
      testDataWithSelections.data[12].selected = true;
      testDataWithSelections.data[23].selected = true;
    });

    it('will dismiss notifications in bulk', async () => {
      const wrapper = mount(NotificationsList, opts);
      await wrapper.setProps({ notifications: testDataWithSelections });

      await wrapper.get(BulkDismiss).trigger('click');
      expect(wrapper.emitted('dismiss')).toEqual([[selectedNotifications]]);
    });

    it('will undismiss notifications in bulk', async () => {
      const wrapper = mount(NotificationsList, opts);
      await wrapper.setProps({ notifications: testDataWithSelections });

      await wrapper.get(BulkUndismiss).trigger('click');
      expect(wrapper.emitted('undismiss')).toEqual([[selectedNotifications]]);
    });

    it('will delete notifications in bulk', async () => {
      const wrapper = mount(NotificationsList, opts);
      await wrapper.setProps({ notifications: testDataWithSelections });

      await wrapper.get(BulkDelete).trigger('click');
      expect(wrapper.emitted('delete')).toEqual([[selectedNotifications]]);
    });
  });
});
