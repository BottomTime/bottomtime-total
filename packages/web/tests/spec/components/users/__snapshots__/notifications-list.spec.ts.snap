// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsList component will render a list of notifications 1`] = `"Showing50of137notifications."`;
